### DISCUSSION TITLE
Do you have Honest feedback? Help us improve

### DISCUSSION THREAD

**Discussion by TomasVotruba** [rectorphp/rector#6411] [General]
Hi,

this is place to share any feedback that you don't want to make a new issue for.
It should focus [on little papercut annoyances](https://twitter.com/migueldeicaza/status/1392154167486160896), that make you little angry inside on daily basis.

We want to get rid of these details, so using Rector is fun and relaxing :slightly_smiling_face: 

Please, be constructive and suggest a solution, so we know what you expected and which way to go :+1: 

Thank you

**Comment by Wirone** [comment]
@TomasVotruba I think this will be good place for my case. Every time I want to contribute to Rector I am struggling with runtime environment.

- Repo does not have encapsulated runtime (Docker, Vagrant, anything) for running, it assumes all required dependencies are directly on developer's computer which is not fitting in 2021 😉 
- Repo does not have technical guide for contribution (what commands should be executed in order to get complete runtime?), there is only "[code of conduct](https://github.com/rector/rector-src/blob/main/CONTRIBUTING.md)" for contributors.
- Even after clean `git clone` there is an error:
  ```
  ❯ composer install

  No composer.lock file present. Updating dependencies to latest instead of installing from lock file. See https://getcomposer.org/install for more information.
  Loading composer repositories with package information
  Updating dependencies
  Your requirements could not be resolved to an installable set of packages.

    Problem 1
      - Root composer.json requires rector/rector-generator ^0.4 -> satisfiable by rector/rector-generator[0.4.0].
      - rector/rector-generator 0.4.0 requires rector/rector-src ^0.11.57 -> satisfiable by rector/rector-src[dev-main, 0.11.57, 0.11.58, 0.11.x-dev (alias of dev-main)] from composer repo (https://repo.packagist.org) but rector/rector-src is the root package and cannot be modified. See https://getcomposer.org/dep-on-root for details and assistance.
  ```
  I always end up with removing `rector/rector-generator` from `composer.json` in order to work with the code.
- after `git pull` and `composer install` (after hacking) I had errors `_PHPStan_76800bfb5\Nette\DI\ServiceCreationException: Service of type Rector\PHPStanRules\NodeAnalyzer\SymfonyConfigRectorValueObjectResolver: Service of type PhpParser\NodeFinder needed by $nodeFinder in __construct() not found. Did you add it to configuration file?`, PHPUnit started working only after fresh `git clone` (removing `vendor` and `tmp` dirs did not work). I had several different issues like this and it's frustrating to waste time on fixing runtime.

Probably you have workflow so much established that many things are obvious for you, but for new people whom would like to help it can be annoying and deterrent.

And this is from person who worked with ECS and Rector in the past and thinks about self as a experienced coder, so probably less experienced people would struggle even more 😉

**Reply by TomasVotruba** [reply]
:+1: 

Thank you for feedback. Let's take it one by one :)

1. We're not using Docker for development, only because no long-term contributor needed it so far. Could you add your Docker setup that works for you?

2. The link is mentioned in the `README.md` and goes to: https://github.com/rectorphp/rector-src/blob/main/CONTRIBUTING.md Could you add it to the place where you expect to find it?

3. That's a new bug. I tried to fix it. Could you re-try?

4. I'm trying to replicate this one, but failed. It works for me. Could you share repository with same steps in CI Github Action? So we can see what is wrong there.

<br>

One more thing, if you see place for improvement, please send pull-request with you idea. We're open to improvement and seing PR is usually much easier to process than issues and long-dragging discussions about them. See the last PR you've send and we've just merged :+1:

**Reply by Wirone** [reply]
1. If you're open for it, I can send minimal Docker setup that I use for running Rector locally 👍 
2. I DID find `CONTRIBUTING.md` but it contains only information on how PR should be prepared. It does not help with preparing local environment for working with the code: `git clone`, `composer install` and so on - everything that should be done in order to get it running (so mentioned `composer complete-check` can be actually executed).
3. Looks like it works on current `main` 👍 
4. I had fork of `rector-src` cloned locally some time ago, today I did `git pull upstream main`, `composer install` and so on and when I wanted to run tests I had those errors. Clearing `tmp` etc did not work, so I cloned repo from scratch. I don't think I can reproduce it right now. Let's hope it will work from now 😉

**Reply by TomasVotruba** [reply]
1. Got for it :+1: 

2. Also, send PR with the content that would help you to find there.
 
3. Thanks, I'm glad it works! :hugs: 

4. Good :) If it will break again, please send the reproducible repository, so we can be sure we fixed it :+1:

**Reply by Wirone** [reply]
Fixed 1 & 2 with rectorphp/rector-src#998 👍

**Comment by 7thstorm** [comment]
First of all, this has been a life saver. We have been putting off upgrades and now, we're going through it, but with much less pain than anticipated. Want to thank everyone in this project.

A few possible improvements that should be easy to implement:

1. If file or path does not exist, do not show `[OK] Rector is done!`
It has taken hours trying to figure out why rector was saying OK, when the files clearly had issues. I misspelled the path, so the file did not exist. I hope this is taken into consideration, or if I'm missing a parameter, pls let me know


2. Provide an option to save a copy of the original file before changing the file. example: filename.php.rector.
for those running rector on hundreds of files, it becomes harder even if we search by recent change date. Hope this is easy to implement and may help a lot of folks

**Reply by Wirone** [reply]
1. Seems like a bug, I can't test it right now but it surely should report wrongly configured paths.
2. Just use VCS and/or `--dry-run` CLI option 🙂.
